### ISSUE TITLE
Invalid nested parameter format (2)

### ISSUE THREAD

**Issue by HolQue** [test-fullautomation/python-jsonpreprocessor#118]
The JSON code

```
"list_param" : ["A", "B", "C"],
"val1"       : ${list_param[0]}
```

causes the following error message:

`'Invalid nested parameter format: ${list_param[0]}'`

The wording is wrong: The expression '`${list_param[0]}`' does not contain any nested parameters.

(This would be a nesting of parameters: '`${list_param[${index}]}`')

The meaning is valid - but unfortunately not helpful for users. They need to be informed about the reason: Why is the format invalid?

Expected is an error message telling, that it's not allowed to put the index inside the curly brackets. Indices have to be placed outside.

**Comment by namsonx** [comment]
Hello Holger,

In your example case, the nested parameter should be 
```
${list_param}[0]
```
instead of 
```
${list_param[0]}
```

Thank you,
Son

**Comment by HolQue** [comment]
Hi Son,

?????????????????????

What is going on here?

At first: It's a "bad base" test.
At second: I clearly wrote down my expectation: "Expected is an error message telling, that it's not allowed to put the index inside the curly brackets. Indices have to be placed outside."

What is your problem?

**Comment by namsonx** [comment]
Hello Holger,

Sorry, I was not aware of this is a bad case because I didn't read your comment carefully.
There are many different ways making a nested parameter format is invalid. It will take a lot of effort to write a code to analyze for every single invalid case, then give a hint in error message.

I suggest we just provide the rule of how to use nested parameter whenever wrong format detection instead of detecting every invalid case.

Thank you,
Son

**Comment by HolQue** [comment]
Hi Son,

I agree: it will not be easy ;-)

Maybe it helps to inverse the logic: We define syntax rules. This causes code that is able to check the JSON content against violations of these rules. Otherwise we would not be able to compute this content. In my opinion you need for the check of every rule a separate method in your code. Every of these methods returns the information if the rule is violated or not. With this strategy it should be possible to provide meaningful error messages.

Mentioning these rules only in documentation is not enough. If these rules are not handled in the code also, the entire application is not able to work. Lots of error messages are currently so much misleading. And some of them are invalid also. This urgently has to be improved.

In

https://github.com/test-fullautomation/python-jsonpreprocessor/blob/develop/test/testconfig/Checklist.txt

I already wrote down all possible rules and strategies, that came into my mind. Please try to get ideas out of it to improve the error messages of JsonPreprocessor.

**Comment by test-fullautomation** [comment]
@namsonx : What is the status of this issue?

**Comment by namsonx** [comment]
Hello Holger, Hello Thomas,

This ticket was solved with latest commit on stabi branch.
The error message log in this case is ```Invalid syntax: Found index inside curly brackets in line '"val1"       : ${list_param[0]}'. Indices in square brackets have to be placed outside the curly brackets.```

Thank you,
Son

**Comment by test-fullautomation** [comment]
Hi Son,
looks good for me!
Thank you,
 Thomas

**Comment by HolQue** [comment]
Retest successful, Issue can be closed.

**Comment by namsonx** [comment]
merged to develop branch

**Comment by test-fullautomation** [comment]
solved with version 0.10.0
